index.bs

```diff
@@ -795,6 +743,562 @@ the following list:
+
+  : {{WebSocketOpenInfo/extensions}}
+  :: The [=extensions in use=] for the connection.
+
+  : {{WebSocketOpenInfo/protocol}}
+  :: The [=subprotocol in use=] for the connection.
+
+ : <code>|socket|.{{WebSocketStream/closed}}</code>
+ :: A {{promise}} which resolves when the connection is closed. If the connection did not close
+  [=cleanly=] then the promise is rejected with a {{WebSocketError}}. When the connection closes
+  [=cleanly=] the promise is fulfilled with an object with properties
+  {{WebSocketCloseInfo/closeCode}} and {{WebSocketCloseInfo/reason}}, giving [=the WebSocket
+  connection close code=] and [=the WebSocket connection close reason=] that were supplied by the
+  server.
+
+ : <code>|socket|.{{WebSocketStream/close()}}</code>
+ :: Close the connection, optionally supplying an object with {{WebSocketCloseInfo/closeCode}} and
+  {{WebSocketCloseInfo/reason}} properties to indicate [=the WebSocket connection close code=] and
+  [=the WebSocket connection close reason=] that will be sent to the remote server. If the handshake
+  is still in progress, then it will be aborted and {{WebSocketCloseInfo/closeCode}} and
+  {{WebSocketCloseInfo/reason}} will be ignored.
+</dl>
+
+<div algorithm>
+
+ The <dfn constructor for="WebSocketStream" data-lt="WebSocketStream(url, options)"><code>new
+ WebSocketStream(|url|, |options|)</code></dfn> constructor steps are:
+
+ 1. Let |baseURL| be [=this=]'s [=relevant settings object=]'s [=API base URL=].
+ 1. Let |urlRecord| be the result of [=get a URL record|getting a URL record=] given |url| and
+    |baseURL|.
+ 1. Let |protocols| be |options|["{{WebSocketStreamOptions/protocols}}"] if it [=map/exists=],
+    otherwise an empty sequence.
+ 1. If any of the values in |protocols| occur more than once or otherwise fail to match the
+    requirements for elements that comprise the value of
+    \`<a http-header>`Sec-WebSocket-Protocol`</a>\` fields as defined by The WebSocket protocol,
+    then throw a "{{SyntaxError}}" {{DOMException}}.  [[!WSP]]
+ 1. Set [=this=]'s [=WebSocketStream/url=] to |urlRecord|.
+ 1. Set [=this=]'s [=WebSocketStream/opened promise=] and [=WebSocketStream/closed promise=] to new
+    promises.
+ 1. Apply backpressure to the WebSocket.
+
+  Note: This means that messages won't be read until the application is ready for them.
+
+ 1. If |options|["{{WebSocketStreamOptions/signal}}"] [=map/exists=],
+  1. Let |signal| be |options|["{{WebSocketStreamOptions/signal}}"].
+  1. If |signal| is [=AbortSignal/aborted=], then reject [=this=]'s [=WebSocketStream/opened
+     promise=] and [=WebSocketStream/closed promise=] with |signal|'s [=abort reason=] and return.
+  1. [=AbortSignal/add|Add the following abort steps=] to |signal|:
+   1. If the WebSocket connection is not yet [=established=] [[!WSP]]
+    1. [=Fail the WebSocket connection=] and set [=this=]'s [=WebSocketStream/ready state=] to
```

Comment:
This could probably be split into two steps.

Reply:
Done.